jina/executors/indexers/cache.py

```diff
@@ -24,33 +24,58 @@ def post_init(self):
 class DocIDCache(BaseCache):
     """Store doc ids in a int64 set and persistent it to a numpy array """
 
+    # TODO temp data structure
+    # TODO can we use BinaryPbInd inside this?
+    store = {}
+
     def __init__(self, index_filename: str = None, *args, **kwargs):
         if not index_filename:
             # create a new temp file if not exist
             index_filename = tempfile.NamedTemporaryFile(delete=False).name
         super().__init__(index_filename, *args, **kwargs)
 
     def add(self, doc_id: UniqueId, *args, **kwargs):
-        d_id = int(doc_id)
-        self.query_handler.add(d_id)
         self._size += 1
-        self.write_handler.write(np.int64(d_id).tobytes())
+        self.store[doc_id] = kwargs.get(DATA_FIELD)
+
+    @deprecated_alias(doc_id='data')
+    def query(self, data: UniqueId, *args, **kwargs) -> Optional[bool]:
+        """
+        @param data: either the `id` or the `content_hash` of a `Document`
+        """
+        # print(f'query = {data}, self.store = {self.store}')
+        status = (data in self.store.values()) or None
+        print(f'query = {data}. status = {status}')
+        return status
 
-    def query(self, doc_id: UniqueId, *args, **kwargs) -> Optional[bool]:
-        d_id = int(doc_id)
-        return (d_id in self.query_handler) or None
-
-    def get_query_handler(self):
-        with open(self.index_abspath, 'rb') as fp:
-            return set(np.frombuffer(fp.read(), dtype=np.int64))
-
-    @cached_property
-    def null_query_handler(self):
-        """The empty query handler when :meth:`get_query_handler` fails"""
-        return set()
+    def get_create_handler(self):
+        pass
+
+    def update(self, keys: Iterator[int], values: Iterator[bytes], *args, **kwargs):
+        """
+        :param keys: list of Document.id
+        :param values: list of either `id` or `content_hash` of :class:`Document"""
+        missed = check_keys_exist(keys, self.store.keys())
+        if missed:
+            raise KeyError(f'Keys {missed} were not found. No operation performed...')
+
+        for key, value in zip(keys, values):
+            self.store[key] = value
+
+    def delete(self, keys: Iterator[int], *args, **kwargs):
+        """
+        :param keys: list of Document.id
+        """
+        missed = check_keys_exist(keys, self.store.keys())
+        if missed:
+            raise KeyError(f'Keys {missed} were not found. No operation performed...')
+
+        for key in keys:
+            del self.store[key]
+            self._size -= 1
 
     def get_add_handler(self):
-        return open(self.index_abspath, 'ab')
+        pass
```

Comment:
who will implement these?

Reply:
That's one of the things to discuss still. What should the internal data structure be? Can I have a `BinaryPbIndexer` internally? How do I assign it a filename based on the filename that the `DocIDCache` is assigned?

Reply:
I do not think a `BinaryPbIndexer` is the right one, since you do not store `Protobuf` documents. It is a KVIndexer and you can just write to a file as it was done previously right? The only difference is what will be written there or read

Reply:
Previously it was only one array. Now I need a KV indexer, yes. But do I implement another one just for this?

Reply:
Also I'm still not sure whether the field should be set in the Driver or in the DocIDCache. If we don't set it in DocIDCache, we duplicate data. However, if we don't set it in Driver we break the API of the CRUD methods (because we would need to pass the entire document to a query (not just a field)

Reply:
you could have it in the executor and consulted from the Driver too right?

Reply:
> Previously it was only one array. Now I need a KV indexer, yes. But do I implement another one just for this?

The point is, by inheritance, DocIDCache is a `KVIndexer` and it was implemented as this, so yes we have are indeed implementing a new one

Reply:
But the internal data structure should also be completely new? Just using a `dict` is probably not efficient enough, right? Before we were only storing one np array of the ids. Now we need kv of `docid -> doc.content_hash`

Reply:
It could be two arrays, mapped from one to the other by positional index

Reply:
seems a good idea